### PR TITLE
Feat/ot170 23 endpoint-user-registration

### PR DESCRIPTION
**Pull request by pabloayalaalvarez**
Summary:

Create registration
Create authentication_controller
The user is added to the database
Add gems jwt and brcypt

### CODE REVIEW

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -1,4 +1,9 @@
-# frozen_string_literal: true
-
```
Comment (FernandoCiballero) [inline app/controllers/application_controller.rb:2]: Be careful not to delete this.

File: Gemfile [hunk 2]
```diff
@@ -8,8 +8,12 @@ ruby '3.0.2'
 gem 'rails', '~> 6.1.5'
 
 gem 'bootsnap', '>= 1.4.4', require: false # https://rubygems.org/gems/bootsnap
+gem 'bcrypt'
+gem 'discard', '~> 1.2'
 gem 'dotenv-rails', '~> 2.7', '>= 2.7.6' # https://rubygems.org/gems/dotenv-rails
+gem "image_processing", ">= 1.2"
```
Comment (FernandoCiballero) [inline Gemfile:14]: Why did you decide to add image_processing?

Comment (pabloayalaalvarez) [reply Gemfile:14]: I have understood that it is necessary for image analysis and processing.

File: Gemfile [hunk 3]
```diff
@@ -8,8 +8,12 @@ ruby '3.0.2'
 gem 'rails', '~> 6.1.5'
 
 gem 'bootsnap', '>= 1.4.4', require: false # https://rubygems.org/gems/bootsnap
+gem 'bcrypt'
+gem 'discard', '~> 1.2'
```
Comment (FernandoCiballero) [inline Gemfile:12]: Please, keep the same format with the version of the gem and github url

File: Gemfile [hunk 4]
```diff
@@ -8,8 +8,12 @@ ruby '3.0.2'
...
 gem 'dotenv-rails', '~> 2.7', '>= 2.7.6' # https://rubygems.org/gems/dotenv-rails
+gem "image_processing", ">= 1.2"
 gem 'jsonapi-serializer', '~> 2.2' # https://rubygems.org/gems/jsonapi-serializer
+gem 'jwt'
```
Comment (FernandoCiballero) [inline Gemfile:16]: Idem as above.

File: Gemfile (hunk 2 above)
Comment (FernandoCiballero) [reply Gemfile:14]: Ok, I understand but in this case, we have chosen to use Active Storage for file processing.

**Review by FernandoCiballero — COMMENTED** [review]
Geat  Job Pablo!!! 👏  left some comments.

File: app/controllers/application_controller.rb [hunk 5]
```diff
@@ -1,4 +1,9 @@
...
 class ApplicationController < ActionController::API
+  before_action :authenticate_request
+  attr_reader :current_user
+
+  def authenticate_request
+    @current_user = AuthorizeApiRequest.call(request.headers).result
```
Comment (FernandoCiballero) [inline app/controllers/application_controller.rb:6]: Where is located this `AuthorizeApiRequest` class?

Comment (pabloayalaalvarez) [reply app/controllers/application_controller.rb:6]: That's the part of the token I'm working on. I'm having some trouble generating it. I thought I left that part out.

File: Gemfile [hunk 6]
```diff
@@ -8,8 +8,11 @@ ruby '3.0.2'
 gem 'rails', '~> 6.1.5'
 
 gem 'bootsnap', '>= 1.4.4', require: false # https://rubygems.org/gems/bootsnap
+gem 'bcrypt'
```
Comment (DevChristians) [inline Gemfile:11]: add commentary # repo gem

File: Gemfile [hunk 7]
```diff
@@ -8,8 +8,11 @@ ruby '3.0.2'
...
+gem 'discard', '~> 1.2'
 gem 'dotenv-rails', '~> 2.7', '>= 2.7.6' # https://rubygems.org/gems/dotenv-rails
 gem 'jsonapi-serializer', '~> 2.2' # https://rubygems.org/gems/jsonapi-serializer
+gem 'jwt'
```
Comment (DevChristians) [inline Gemfile:15]: add comment # repo link

**Review by DevChristians — CHANGES_REQUESTED** [review]

File: app/controllers/application_controller.rb [hunk 8]
```diff
@@ -1,4 +1,18 @@
-# frozen_string_literal: true
-
 class ApplicationController < ActionController::API
+  before_action :authenticate!
```
Comment (Ezequiel-Padilla) [inline app/controllers/application_controller.rb:2]: why do we need the before action here?

File: app/controllers/application_controller.rb [hunk 9]
```diff
@@ -1,4 +1,18 @@
...
+    header = header.split(' ').last if header
+    begin
+      @decoded = JsonWebToken.decode(header)
+      @current_user = User.find(@decoded[:user_id])
+    rescue ActiveRecord::RecordNotFound => e
+      render json: { errors: e.message }, status: :unauthorized
+    rescue JWT::DecodeError => e
+      render json: { errors: e.message }, status: :unauthorized
+    end
+  end
```
Comment (Ezequiel-Padilla) [inline app/controllers/application_controller.rb:17]: run rubocop here

File: config/environments/production.rb [hunk 10]
```diff
@@ -33,6 +33,9 @@
   # Store uploaded files on the local file system (see config/storage.yml for options).
   config.active_storage.service = :local
 
+  # Store files on Amazon S3.
+  config.active_storage.service = :amazon
+
```
Comment (Ezequiel-Padilla) [inline config/environments/production.rb:38]: because we added this

File: config/routes.rb [hunk 11]
```diff
@@ -1,7 +1,10 @@
 Rails.application.routes.draw do
-  resources :testimonials
+
```
Comment (Ezequiel-Padilla) [inline config/routes.rb:2]: remove this line

File: spec/factories/users.rb [hunk 12]
```diff
@@ -0,0 +1,9 @@
+FactoryBot.define do
+  factory :user do
+    firstName { "MyString" }
+    lastName { "MyString" }
+    email { "MyString" }
+    password_digest { "MyString" }
+    photo { "MyString" }
+  end
```
Comment (Ezequiel-Padilla) [inline spec/factories/users.rb:8]: use faker

File: spec/models/user_spec.rb [hunk 13]
```diff
@@ -0,0 +1,5 @@
+require 'rails_helper'
+
+RSpec.describe User, type: :model do
+  pending "add some examples to (or delete) #{__FILE__}"
+end
```
Comment (Ezequiel-Padilla) [inline spec/models/user_spec.rb:5]: if we add endpoint tests let's try to add the model tests

File: Gemfile (hunk 7 above)
Comment (FernandoCiballero) [reply Gemfile:14]: And don't forget to add the gem version.

File: app/controllers/api/v1/users_controller.rb [hunk 14]
```diff
@@ -3,6 +3,31 @@
 module Api
   module V1
     class UsersController < ApplicationController
+      skip_before_action :authenticate!, only: :create
```
Comment (FernandoCiballero) [inline app/controllers/api/v1/users_controller.rb:6]: Please check if you want this `skip_before_action`.

File: app/controllers/api/v1/users_controller.rb [hunk 15]
```diff
@@ -3,6 +3,31 @@
...
+        users = User.all
+        render json: users.as_json(json_options)
+      end
+
+      def create
+        user = User.new(user_params)
+        if user.save
+          render json: user.as_json(json_options)
+        else
+          render json: { status: :bad, errors: user.errors.messages }
```
Comment (FernandoCiballero) [inline app/controllers/api/v1/users_controller.rb:18]: Please check the [http status codes](http://www.railsstatuscodes.com/).

File: app/controllers/api/v1/users_controller.rb [hunk 16]
```diff
@@ -3,6 +3,31 @@
 module Api
   module V1
     class UsersController < ApplicationController
+      skip_before_action :authenticate!, only: :create
+
+      def index
+        users = User.all
+        render json: users.as_json(json_options)
+      end
```
Comment (FernandoCiballero) [inline app/controllers/api/v1/users_controller.rb:11]: Maybe this action is not required in this ticket.

File: app/controllers/api/v1/users_controller.rb [hunk 17]
```diff
@@ -3,6 +3,31 @@
...
+
+      def create
+        user = User.new(user_params)
+        if user.save
+          render json: user.as_json(json_options)
```
Comment (FernandoCiballero) [inline app/controllers/api/v1/users_controller.rb:16]: You can use a [serializer](https://github.com/jsonapi-serializer/jsonapi-serializer) to show the created user.

File: config/routes.rb [hunk 18]
```diff
@@ -1,6 +1,9 @@
 Rails.application.routes.draw do
   namespace :api, defaults: { format: 'json' } do
     namespace :v1 do
+      post :users, to: 'users#create', path: 'auth/register'
+      get :users, to: 'users#index'
+      post :login, to: 'authentication#create'
```
Comment (FernandoCiballero) [inline config/routes.rb:6]: Please check if this routes are in the ticket requirements.

File: db/migrate/20220326181847_add_jti_to_user.rb [hunk 19]
```diff
@@ -0,0 +1,6 @@
+class AddJtiToUser < ActiveRecord::Migration[6.1]
+  def change
+    add_column :users, :jti, :string
```
Comment (FernandoCiballero) [inline db/migrate/20220326181847_add_jti_to_user.rb:3]: Why do we need this field in the user table?

**Pull request closed without merging**